### PR TITLE
SI-7149 Use a WeakHashSet for type uniqueness

### PR DESCRIPTION
**Pull request by JamesIry**
We have an implementation of WeakHashSet in the compiler but it never cleans up its WeakReferences. The first commit in this PR replaces the implementation with one that does clean up. Along the way it modifies the interface to follow the standard scala.collection.mutable.Set interface so that it can eventually be moved into the main library should we choose to do so.

Currently type uniqueness is done via a HashSet[Type], but
that means the Types live through an entire compile session, even
ones that are used once. The result is a huge amount of unnecessarily
retained memory. This commit uses a WeakHashSet instead so that Types
and their WeakReferences are cleaned up when no longer in use.

Finally, perRunCaches is also updated to use the WeakHashSet instead of a HashSet of WeakReferences because the former was never cleaning up unused WeakReferences.

### CODE REVIEW

**Comment by JamesIry** [comment]
Review @gkossakowski.

**Comment by JamesIry** [comment]
Cleaned up several typos in comments.

**Comment by gkossakowski** [comment]
> Currently type uniqueness is done via a HashSet of WeakReferences, but
> that means the WeakReferences live through an entire compile session.
> Another commit uses the WeakHashSet instead so that the WeakReferences are
> cleaned up when no longer in use.

I thought we were not using any form of WeakReferences in uniques set thus we were accumulating those types for the entire compiler run. If you look at the diff in 4eb4aa5 you'll see that there are no WeakReferences involved there. Am I missing something?

**Comment by JamesIry** [comment]
You're right, my commit message is wrong.

**Comment by gkossakowski** [comment]
@JamesIry: could you update the commit and PR message?

**Comment by JamesIry** [comment]
Cleaned up the commit message and added a comment to the WeakHashSet's hash avalanche mechanism.

**Comment by gkossakowski** [comment]
LGTM.

Great work!

I did memory profiling of this change. I'm comparing b88801f (parent commit of this PR) and 29babb5 (last commit in this PR). I used YourKit and [custom probe](https://gist.github.com/gkossakowski/5701350) to collect memory snapshot after different phases. See YourKit [documentation](http://www.yourkit.com/docs/java/help/probes.jsp) for detailed explanation of probes.

I compiled Scala library with probes enabled like this:

```
ant -Djvm.opts="-agentpath:/Applications/YourKit.app/bin/mac/libyjpagent.jnilib=builtinprobes=none,probeclasspath=/Users/grek/scala/scala-master/probe-classes,probe=MemoryProbe" quick.lib
```

Now the numbers (for compiling the Scala library)

```
Total retained memory after phase (in mb):
=========================================
Phase       Before      After   Lost
typer       173         136     37 (21%)
erasure     264         178     86 (32%)
cleanup     314         225     89 (28%)

Number of retained :: objects after each phase (in k):
=====================================================
Phase       Before      After   Lost
typer       808         502     306 (37%)
erasure     1432        766     666 (46%)
cleanup     1737        1044    693 (39%)

Number of retained TypeHistory objects after each phase (in k):
==============================================================
Phase       Before      After   Lost
typer       202         123     79  (39%)
erasure     473         259     214 (45%)
cleanup     605         378     227 (37%)


Number of retained ClassArgsTypeRef objects after each phase (in k):
===================================================================
Phase       Before      After   Lost
typer       226         88      138 (61%)
erasure     396         78      318 (80%)
cleanup     419         88      331 (78%)
```

Great work, again!

**Comment by xeno-by** [comment]
You got to be kidding me!!

**Comment by viktorklang** [comment]
FANTASTIC

**Comment by ihji** [comment]
Wow, the benchmark result is TOTALLY awesome. Great work!

**Comment by fernandoracca** [comment]
excellent job!

**Comment by jrudolph** [comment]
:+1:

File: src/reflect/scala/reflect/internal/util/WeakHashSet.scala [hunk 1]
```diff
@@ -1,59 +1,430 @@
...
+import java.lang.ref.{WeakReference, ReferenceQueue}
+import scala.annotation.tailrec
 import scala.collection.generic.Clearable
-import scala.runtime.AbstractFunction1
+import scala.collection.mutable.{Set => mSet}
 
-/** A bare-bones implementation of a mutable `Set` that uses weak references
- *  to hold the elements.
+/**
+ * A HashSet where the elements are stored weakly. Elements in this set are elligible for GC if no other
```
Comment (etorreborre) [inline src/reflect/scala/reflect/internal/util/WeakHashSet.scala:10]: typo: eligible

File: src/reflect/scala/reflect/internal/util/WeakHashSet.scala [hunk 2]
```diff
@@ -1,59 +1,430 @@
...
+ * hard references are associated with them. Its primary use case is as a canonical reference
+ * identity holder (aka "hash-consing") via findEntryOrUpdate
  *
- *  This implementation offers only add/remove/test operations,
- *  therefore it does not fulfill the contract of Scala collection sets.
+ * This Set implementation cannot hold null. Any attempt to put a null in it will result in a NullPointerException
+ *
+ * This set implmeentation is not in general thread safe without external concurrency control. However it behaves
```
Comment (etorreborre) [inline src/reflect/scala/reflect/internal/util/WeakHashSet.scala:16]: typo: implementation

**Pull request merged**